cmd/clusterctl/pkg/client/cluster/mover.go

```diff
@@ -0,0 +1,112 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ObjectMover defines methods for moving Cluster API objects to another management cluster.
+type ObjectMover interface {
+
+	// Move moves  all the Cluster API objects existing in a namespace (or from all the namespaces if empty) to a target management cluster
+	Move(namespace string, toCluster Client) error
+}
+
+// MachineWaiter defines a function that waits for a machine to be reconciled and linked with the corresponding node.
+type MachineWaiter func(proxy Proxy, machineKey client.ObjectKey) error
+
+// objectMover implements the ObjectMover interface.
+type objectMover struct {
+	fromProxy     Proxy
+	machineWaiter MachineWaiter
+	log           logr.Logger
+}
+
+// ensure objectMover implements the ObjectMover interface.
+var _ ObjectMover = &objectMover{}
+
+func (o *objectMover) Move(namespace string, toCluster Client) error {
+	objectGraph := newObjectGraph(o.fromProxy, toCluster.Proxy(), o.machineWaiter, o.log)
+
+	//TODO: implement preflight checks ensuring the target cluster has all the required providers in place
+
+	// Gets all the types defines by the CRDs installed by clusterctl plus the ConfigMap/Secret core types.
+	types, err := objectGraph.getDiscoveryTypes()
+	if err != nil {
+		return err
+	}
+
+	// Discovery the object graph for the selected types:
+	// - Nodes are defined the Kubernetes objects (Clusters, Machines etc.) identified during the discovery process.
+	// - Edges are derived by the OwnerReferences between nodes.
+	if err := objectGraph.Discovery(namespace, types); err != nil {
+		return err
+	}
+
+	// Identify soft relations between nodes such as secrets linked to the cluster by a naming convention (without any explicit OwnerReference).
+	objectGraph.SetSoftDependents()
+
+	//TODO: add a preflight check ensuring all the Clusters/Machines are already provisioned.
+	//TODO: consider if to add additional preflight checks ensuring the object graph is complete (no virtual nodes left)
+	//TODO: consider if to add additional preflight checks ensuring there are no nodes shared across clusters (or implement support for shared nodes, potentially required by CAPV)
+
+	// Move the objects to the target cluster.
+	if err := objectGraph.Move(); err != nil {
```

Comment:
Same comment as elsewhere - would it make sense to keep `objectGraph` as just a graph, and implement all the moving logic in here in `objectMover`?

Reply:
done